pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,329 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: jefftk@google.com (Jeff Kaufman)
+
+#include "pagespeed/system/controller_manager.h"
+
+#include <signal.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <cerrno>
+#include <cstddef>
+#include <cstdlib>
+
+#include "base/logging.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/system/system_rewrite_options.h"
+
+namespace net_instaweb {
+
+
+ControllerManager::ControllerManager(
+    SystemRewriteDriverFactory* factory,
+    ThreadSystem* thread_system,
+    FileSystem* file_system,
+    Timer* timer,
+    MessageHandler* handler) : factory_(factory),
+                               thread_system_(thread_system),
+                               file_system_(file_system),
+                               timer_(timer),
+                               handler_(handler),
+                               babysitter_pid_(-1) { }
+
+void ControllerManager::SetUpSignalHandlers() {
+  // We need to clear inherited signal handlers. There's no portable way to get
+  // a list of all possible signals, and they're not even guaranteed to be in
+  // order.  But NSIG is usually defined these days, and if it is then we just
+  // want ascending numbers up to to NSIG.
+  for (int i = 0; i < NSIG; i++) {
+    signal(i, SIG_DFL);
+  }
+}
+
+ControllerManager::ProcessDeathWatcherThread::ProcessDeathWatcherThread(
+    ThreadSystem* thread_system,
+    Timer* timer,
+    MessageHandler* handler,
+    pid_t master_pid,
+    pid_t babysitter_pid) : Thread(thread_system,
+                                   "master process death watcher",
+                                   ThreadSystem::kDetached),
+                            timer_(timer),
+                            handler_(handler),
+                            master_pid_(master_pid),
+                            babysitter_pid_(babysitter_pid) {
+}
+
+bool IsProcessRunning(pid_t pid) {
+  // "If sig is 0, then no signal is sent, but error checking is still
+  //  performed; this can be used to check for the existence of a process ID
+  //  or process group ID."
+  return kill(pid, 0) == 0;
+}
+
+void ControllerManager::ProcessDeathWatcherThread::Run() {
+  handler_->Message(
+      kInfo, "Watching master (%d) and babysitter (%d) to exit if they do.",
+      master_pid_, babysitter_pid_);
+  while (IsProcessRunning(master_pid_) &&
+         IsProcessRunning(babysitter_pid_)) {
+    timer_->SleepUs(30 * 1000 /* 0.03s */);
```

Comment:
You can't wait on a random pid, and because the root process daemonizes, it might as well be random.

We might be able to do something more efficient with pipes, where I keep open pipes between these, and then if the other end dies I'll see something happen.  I'll look into that.